azurerm/internal/services/cognitive/cognitive_account_customer_managed_key_resource.go

```diff
@@ -0,0 +1,253 @@
+
+	id, err := parse.AccountID(d.Get("cognitive_account_id").(string))
+	if err != nil {
+		return err
+	}
+
+	locks.ByName(id.Name, "azurerm_cognitive_account")
+	defer locks.UnlockByName(id.Name, "azurerm_cognitive_account")
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("retrieving %s: %+v", *id, err)
+	}
+
+	if d.IsNewResource() {
+		if resp.Properties != nil && resp.Properties.Encryption != nil {
+			return tf.ImportAsExistsError("azurerm_cognitive_account_customer_managed_key", id.ID())
+		}
+	}
+
+	keySource := cognitiveservices.KeySource(d.Get("key_source").(string))
+	props := cognitiveservices.Account{
+		Properties: &cognitiveservices.AccountProperties{
+			Encryption: &cognitiveservices.Encryption{
+				KeySource: keySource,
+			},
+		},
+	}
+	if keySource == cognitiveservices.KeySourceMicrosoftCognitiveServices {
+		if _, ok := d.GetOk("key_vault_key_id"); ok {
+			return fmt.Errorf("can't specify key_vault_key_id when using Microsoft.CognitiveServices key source")
+		}
+		if _, ok := d.GetOk("identity_client_id"); ok {
+			return fmt.Errorf("can't specify key_name when using Microsoft.CognitiveServices key source")
+		}
+	} else {
+		keyId, err := keyVaultParse.ParseOptionallyVersionedNestedItemID(d.Get("key_vault_key_id").(string))
+		if err != nil {
+			return err
+		}
+		props.Properties.Encryption.KeyVaultProperties = &cognitiveservices.KeyVaultProperties{
+			KeyName:          utils.String(keyId.Name),
+			KeyVersion:       utils.String(keyId.Version),
+			KeyVaultURI:      utils.String(keyId.KeyVaultBaseUrl),
+			IdentityClientID: utils.String(d.Get("identity_client_id").(string)),
+		}
+	}
+
+	if _, err = client.Update(ctx, id.ResourceGroup, id.Name, props); err != nil {
+		return fmt.Errorf("updating %s: %+v", *id, err)
+	}
+
+	stateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{"Accepted"},
+		Target:     []string{"Succeeded"},
+		Refresh:    cognitiveAccountStateRefreshFunc(ctx, client, *id),
+		MinTimeout: 15 * time.Second,
+		Timeout:    d.Timeout(pluginsdk.TimeoutCreate),
+	}
+
+	if _, err = stateConf.WaitForStateContext(ctx); err != nil {
+		return fmt.Errorf("waiting for update of %s: %+v", *id, err)
+	}
+	d.SetId(id.ID())
+
+	return resourceCognitiveAccountCustomerManagedKeyRead(d, meta)
+}
+
+func resourceCognitiveAccountCustomerManagedKeyRead(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Cognitive.AccountsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.AccountID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("retrieving %s: %+v", *id, err)
+	}
+	if resp.Properties == nil && resp.Properties.Encryption == nil {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("cognitive_account_id", id.ID())
+	d.Set("key_source", resp.Properties.Encryption.KeySource)
```

Comment:
Potential crash here, needs nil checking

Reply:
I added nil-check above, so it won't be nil here.